File: 87.Scramble-String.java
// https://leetcode.com/problems/scramble-string/
//
// algorithms
// Medium (31.67%)
// Total Accepted: 91,120
// Total Submissions: 287,720
// beats 100.0% of java submissions

class Solution {
    public boolean isScramble(String s1, String s2) {
        if (s1.equals(s2)) {
            return true;
        }
        int len = s1.length();

        int[] chs = new int[26];
        boolean res = true;
        for (int i = 0; i < len; i++) {
            char ch1 = s1.charAt(i), ch2 = s2.charAt(i);
            chs[ch1 - 'a']++;
            chs[ch2 - 'a']--;
            if (ch1 != ch2) {
                res = false;
            }
        }
        if (res) {
            return true;
        }

        for (int n : chs) {
            if (n != 0) {
                return false;
            }
        }

        for (int i = 1; i < len; i++) {
            if (isScramble(s1.substring(0, i), s2.substring(0, i)) && (isScramble(s1.substring(i), s2.substring(i)))) {
                return true;
            }
            if (isScramble(s1.substring(0, i), s2.substring(len - i))
                    && (isScramble(s1.substring(i), s2.substring(0, len - i)))) {
                return true;
            }
        }

        return false;
    }
}

File: 87.Scramble-String.py
# https://leetcode.com/problems/scramble-string/
#
# algorithms
# Hard (31.83%)
# Total Accepted:    92,440
# Total Submissions: 290,420


class Solution(object):
    def isScramble(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        length = len(s1)

        if s1 == s2:
            return True

        chs = [0] * 26
        for ch in s1:
            chs[ord(ch) - 97] += 1

        for ch in s2:
            chs[ord(ch) - 97] -= 1

        for n in chs:
            if n != 0:
                return False

        for i in xrange(1, length):
            if self.isScramble(s1[i:], s2[i:]) and self.isScramble(s1[:i], s2[:i]):
                return True

            if self.isScramble(s1[:i], s2[length - i:]) and self.isScramble(s1[i:], s2[:length - i]):
                return True

        return False